src/Aspire.Hosting.DevTunnels/DevTunnelResourceBuilderExtensions.cs

```diff
@@ -0,0 +1,489 @@
+                }
+
+                // Remove the localhost version of the tunnel URL that's added by the central endpoint URL logic
+                // HACK: See the NOTE above about potentially handling this more generically in the central endpoint URL logic
+                if (urls.FirstOrDefault(u => string.Equals(u.Endpoint?.EndpointName, DevTunnelPortResource.TunnelEndpointName, StringComparisons.EndpointAnnotationName)
+                                             && string.Equals(new UriBuilder(u.Url).Host, "localhost", StringComparison.OrdinalIgnoreCase)) is { } localhostTunnelUrl)
+                {
+                    urls.Remove(localhostTunnelUrl);
+                }
+
+                // Remove any existing inspect URL
+                if (urls.FirstOrDefault(u => string.Equals(u.DisplayText, "Inspect", StringComparison.OrdinalIgnoreCase)) is { } inspectUrl)
+                {
+                    urls.Remove(inspectUrl);
+                }
+
+                // Add the inspect URL if available
+                var portResource = (DevTunnelPortResource)context.Resource;
+                if (portResource.LastKnownStatus?.PortUri is { } portUri)
+                {
+                    // If tunnel host is sdfdff-3456.usw.devtunnels.ms, the inspect host is sdfdff-3456-inspect.usw.devtunnels.ms
+                    var hostPrefixLength = portUri.Host.IndexOf('.');
+                    var hostPrefix = portUri.Host[..hostPrefixLength];
+                    var hostSuffix = portUri.Host[hostPrefixLength..];
+                    urls.Add(new()
+                    {
+                        Url = new UriBuilder(portUri) { Host = $"{hostPrefix}-inspect{hostSuffix}" }.Uri.ToString(),
+                        DisplayText = "Inspect",
+                        DisplayLocation = UrlDisplayLocation.DetailsOnly
+                    });
+                }
+            })
+            .WithIconName("VirtualNetwork")
+            .WithInitialState(new()
+            {
+                ResourceType = "DevTunnelPort",
+                CreationTimeStamp = DateTime.UtcNow,
+                State = KnownResourceStates.NotStarted,
+                Properties =
+                [
+                    new(CustomResourceKnownProperties.Source, $"{targetResource.Name}/{targetEndpoint.EndpointName}"),
+                    new("TargetResource", targetResource.Name),
+                    new("TargetEndpoint", targetEndpoint.EndpointName),
+                    new("Protocol", portOptions.Protocol),
+                    new("Description", portOptions.Description),
+                    new("Labels", portOptions.Labels is null ? "" : $"{string.Join(", ", portOptions.Labels)}"),
+                ]
+            })
+            .OnInitializeResource((portResource, e, ct) =>
+            {
+                // Setup event subscriptions for the port lifecycle
+                var eventing = e.Services.GetRequiredService<IDistributedApplicationEventing>();
```

Comment:
Why do you subscribe in the callback instead of up front?

Reply:
'cos I forget the eventing type was on the builder